core/queryalgebra/model/src/main/java/org/eclipse/rdf4j/query/algebra/Var.java

```diff
@@ -42,6 +45,12 @@ public Var(String name, Value value) {
 		setValue(value);
 	}
 
+	public Var(Value value) {
+		this(UUID.randomUUID().toString().replace("-", "") + "_anon");
```

Comment:
Let's make sure this uses the same algorithm as `oer.query.parser.sparql.TupleExprBuilder#createAnonVar`:

```
"_anon_" + UUID.randomUUID().toString().replaceAll("-", "_")
```

In fact if we add this as a convenience thing here in the `Var` class, let's also make sure `TupleExprBuilder` reuses it.

Reply:
I'm actually going to redo some of the use of Var and StatementPattern because things aren't working as I want with the last bit of the support for rsx:targetShape. I was actually considering making some more constructors in Var and also a static constant anonymous instance. Also quite annoying that there is no documentation in the Var class!

Reply:
(I think github might be having some issues, because some of my comments are showing up when I look at the diff, but not here - forgive me if I appear to say things twice)

A few more constructors to make (anonymous) Var creation sounds fine. Not sure about a static constant, at least not for general inside the Var class itself - that sounds problematic in the sense that even if it's anonymous, it still has identity. You can't just reuse the same anonymous var everywhere.

As for documentation: I agree that it could be better, but `Var` is a relatively simple container class, anything specific you're missing specs on?

Reply:
Would be nice with some examples of combinations that are legal and what they match to in the query. And simple definitions for all the fields.

Reply:
I presume that anonymous means that it doesn't have a variable name (which makes me wonder why null doesn't suffice). However it can still have a value. Var with variable name and constant value is probably equivalent to a variable bound at query time.

Reply:
To explain: anonymous vars are vars that have no specific _user-defined_ name in the query. In SPARQL this happens, for example, when a query involves a blank node:

```
 SELECT * WHERE {
      [] rdf:type :Foo ;
         rdfs:label ?label .
 }
```

The `[]` in the above query is represented as an anonymous variable in the algebra. However it still needs a separate identity because (as you can see) it needs to appear in two statement patterns - we are executing a join on the anonymous var. And there may be more than one anonymous var in the same query (a second blank node), which means we can't just reuse the same Var instance everywhere.  

Another point where they occur in SPARQL queries is in path expressions. For example this SPARQL query:

```
?start :p1/p2 ?end.
```

internally gets represented by two statement patterns, where the object-var of the first is the subject-var of the second. Again, the variable is anonymous, but its identity is still important as the 'glue-component' in the path expression.

I agree though that there is some ambiguity and lack of documentation in how the Var class handles this. Having to set the the `isAnonymous` flag _in addition_ to assigning it a generated name sounds superfluous.

Reply:
As for why null doesn't suffice: we need to be able to serialize the algebra. And the internal mechanics of the evaluation engine rely on being able to set values for vars by means of using the variable name as a key in a Binding(Set) object.

Reply:
I think I will migrate away from Var and StatementPattern then since my usecase is similar but sufficiently different. I'm using it as a two step process for retrieving the minimal amount of data from the backing store. Essentially I want to express things like:

For a query:
?a rdf:type ex:Company.
?b ex:subsidiary ?c. 

Bound to whatever matches the patterns:
- ?a rdf:type ex:company
- ?b ex:subsidiary ?c

But sometimes it gets more complicated, for some of the SHACL components and I need a pattern like:
- ?b null ?c
Or
- ?a rdf:type null

For the same query. 

These patterns I run against the changeset for the transaction and then use the results to bind the variables in the query to retrieve less data when validating. 

As an example. If I say that every person that has an address they must have max 1 street name. Then we only need to retrieve people and their street names if either a street name is added or a person is added. We can ignore any removals. And we don't need to revalidate all people and their street names, only ones that could actually now violate the rule based on the changes in the transaction.